pkg/deploy/generator/generators.go

```diff
@@ -32,290 +29,163 @@ var (
+
+	rpParametersTemplate(outputFile string) error
+	rpPreDeployParametersTemplate(outputFile string) error
+
+	sharedDevelopmentEnvTemplate(outputFile string) error
+}
+
 type generator struct {
 	production bool
 }
 
-func newGenerator(production bool) *generator {
+func New(production bool) Generator {
 	return &generator{
 		production: production,
 	}
 }
 
-// GenerateRPTemplates generates RP templates for production and development
-// outputs: database-development.json, rp-development.json, rp-production.json
-func GenerateRPTemplates() error {
-	for _, i := range []struct {
-		templateFile string
-		g            *generator
-	}{
-		{
-			templateFile: fileRPDevelopment,
-			g:            newGenerator(false),
-		},
-		{
-			templateFile: FileRPProduction,
-			g:            newGenerator(true),
-		},
-	} {
-		b, err := json.MarshalIndent(i.g.rpTemplate(), "", "    ")
+func (g *generator) Artifacts() error {
+	if g.production {
+		err := g.managedIdentityTemplate(FileRPProductionManagedIdentity)
 		if err != nil {
 			return err
 		}
-
-		// :-(
-		b = bytes.ReplaceAll(b, []byte(tenantIDHack), []byte("[subscription().tenantId]"))
-		if i.g.production {
-			b = bytes.Replace(b, []byte(`"accessPolicies": []`), []byte(`"accessPolicies": "[concat(variables('clustersKeyvaultAccessPolicies'), parameters('extraKeyvaultAccessPolicies'))]"`), 1)
-			b = bytes.Replace(b, []byte(`"accessPolicies": []`), []byte(`"accessPolicies": "[concat(variables('serviceKeyvaultAccessPolicies'), parameters('extraKeyvaultAccessPolicies'))]"`), 1)
+		err = g.rpTemplate(FileRPProduction)
+		if err != nil {
+			return err
 		}
-
-		b = append(b, byte('\n'))
-
-		err = ioutil.WriteFile(i.templateFile, b, 0666)
+		err = g.preDeployTemplates(FileRPProductionPredeploy)
 		if err != nil {
 			return err
 		}
+		err = g.rpParametersTemplate(fileRPProductionParameters)
+		if err != nil {
+			return err
+		}
+		return g.rpPreDeployParametersTemplate(fileRPProductionPredeployParameters)
+	} else {
+		err := g.rpTemplate(fileRPDevelopment)
+		if err != nil {
+			return err
+		}
+		err = g.preDeployTemplates(fileRPDevelopmentPredeploy)
+		if err != nil {
+			return err
+		}
+		err = g.rpParametersTemplate(fileRPProductionParameters)
+		if err != nil {
+			return err
+		}
+		err = g.rpPreDeployParametersTemplate(fileRPProductionPredeployParameters)
+		if err != nil {
+			return err
+		}
+		err = g.databaseTemplate(fileDatabaseDevelopment)
+		if err != nil {
+			return err
+		}
+		return g.sharedDevelopmentEnvTemplate(fileEnvDevelopment)
 	}
 
-	t := &arm.Template{
-		Schema:         "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
-		ContentVersion: "1.0.0.0",
-		Parameters: map[string]*arm.TemplateParameter{
-			"databaseAccountName": {
-				Type: "string",
-			},
-			"databaseName": {
-				Type: "string",
-			},
-		},
-	}
+}
 
-	g := newGenerator(false)
+func (g *generator) managedIdentityTemplate(output string) error {
```

Comment:
from here down is very repetitive, and I don't think the `if g.production`s add value - it's all encoded directly above.  How about two functions, one for parameters and one for templates.  Either their signature is like `writeTemplate(f func() *arm.Template, filename string) error` or it's like `writeTemplate(t *arm.Template, filename string) error`.

Reply:
`if.production` prevents from fool prof errors. 
Functions - could be done, but first `_function` naming need to be decided, as as stated above - I need a suggestion if you don't like current naming :/